### PR TITLE
Add IpAddr common methods

### PR DESCRIPTION
**Pull request by mathphreak**
Per https://github.com/rust-lang/rfcs/pull/1668#issuecomment-230867962 no RFC is needed here.

The generated documentation for these methods is being weird. It shows a deprecation message referencing #27709 for each of them even though two of the referenced methods were stabilized as part of that issue. I don't know how best to address that.

### CODE REVIEW

File: src/libstd/net/ip.rs [hunk 1]
```diff
@@ -59,6 +59,48 @@ pub enum Ipv6MulticastScope {
     Global
 }
 
+impl IpAddr {
+    /// Returns true for the special 'unspecified' address (0.0.0.0 in IPv4, :: in IPv6).
+    pub fn is_unspecified(&self) -> bool {
```
Comment (sfackler) [inline src/libstd/net/ip.rs]: These will all want to be marked unstable.

Comment (mathphreak) [inline src/libstd/net/ip.rs]: @sfackler Should I use `feature="ip"` and `issue="27709"` or should I use a different feature or issue? (#27709 and `feature="ip"` are what the Ipv4/v6Addr convenience methods already use.)

Comment (sfackler) [inline src/libstd/net/ip.rs]: Yeah, I think reusing that issue and feature makes sense, unless @alexcrichton disagrees.

**Comment by alexcrichton** [comment]
Thanks @mathphreak! Perhaps the documentation could also link to the relevant methods on `Ipv4Addr` and `Ipv6Addr` as well? Other than that looks good to me, I've tagged with `T-libs` to ensure this comes up during triage, but I suspect we'll all be ok with it.

cc @rust-lang/libs

**Comment by brson** [comment]
lgtm

File: src/libstd/net/ip.rs [hunk 2]
```diff
@@ -59,6 +59,66 @@ pub enum Ipv6MulticastScope {
     Global
 }
 
+impl IpAddr {
+    /// Returns true for the special 'unspecified' address ([IPv4], [IPv6]).
+    /// [IPv4]: ../../std/net/struct.Ipv4Addr.html#method.is_unspecified
+    /// [IPv6]: ../../std/net/struct.Ipv6Addr.html#method.is_unspecified
+    #[unstable(feature="ip", issue="27709",
+               reason="recently added and depends on unstable Ipv4Addr.is_unspecified()")]
```
Comment (therealbstern) [inline src/libstd/net/ip.rs]: Per [#27709 (comment)](https://github.com/rust-lang/rust/issues/27709#issuecomment-231280999), ipv4Addr.is_unspecified shouldn't be unstable (though this needs a PR from me to mark it as stable).

Comment (mathphreak) [inline src/libstd/net/ip.rs]: @therealbstern I'll keep an eye out for that PR and update this once it gets merged.

**Comment by aturon** [comment]
I'm fine with these changes.

**Comment by alexcrichton** [comment]
Discussed today during libs triage we decided to merge, thanks again for the PR @mathphreak!

**Comment by alexcrichton** [comment]
@bors: retry

* https://github.com/rust-lang/rust/issues/33361

**Comment by alexcrichton** [comment]
@bors: retry

* #34595

**Pull request merged**